stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1056 @@
+                headers_tx.deref_mut(),
+                &rewarded_parent_miner_block_id,
+                &rewarded_miner_block_id,
+                &miner_payout,
+            )?;
+            for user_payout in user_payouts.into_iter() {
+                StacksChainState::insert_matured_child_user_reward(
+                    headers_tx.deref_mut(),
+                    &rewarded_parent_miner_block_id,
+                    &rewarded_miner_block_id,
+                    &user_payout,
+                )?;
+            }
+            StacksChainState::insert_matured_parent_miner_reward(
+                headers_tx.deref_mut(),
+                &rewarded_parent_miner_block_id,
+                &rewarded_miner_block_id,
+                &parent_payout,
+            )?;
+        }
+
+        if applied_epoch_transition {
+            debug!("Block {} applied an epoch transition", &index_block_hash);
+            let sql = "INSERT INTO epoch_transitions (block_id) VALUES (?)";
+            let args: &[&dyn ToSql] = &[&index_block_hash];
+            headers_tx.deref_mut().execute(sql, args)?;
+        }
+
+        debug!(
+            "Advanced to new tip! {}/{}",
+            new_consensus_hash, new_block_hash,
+        );
+        Ok(new_tip_info)
+    }
+
+    /// This function is called in both `append_block` in blocks.rs (follower) and
+    /// `mine_anchored_block` in miner.rs.
+    /// Processes matured miner rewards, alters liquid supply of ustx, processes
+    /// stx lock events, and marks the microblock public key as used
+    /// Returns stx lockup events.
+    pub fn finish_block(
+        clarity_tx: &mut ClarityTx,
+        miner_payouts: Option<&(MinerReward, Vec<MinerReward>, MinerReward, MinerRewardInfo)>,
+    ) -> Result<Vec<StacksTransactionEvent>, ChainstateError> {
+        // add miner payments
+        if let Some((ref miner_reward, ref user_rewards, ref parent_reward, _)) = miner_payouts {
+            // grant in order by miner, then users
+            let matured_ustx = StacksChainState::process_matured_miner_rewards(
+                clarity_tx,
+                miner_reward,
+                user_rewards,
+                parent_reward,
+            )?;
+
+            clarity_tx.increment_ustx_liquid_supply(matured_ustx);
+        }
+
+        // process unlocks
+        let (new_unlocked_ustx, lockup_events) = StacksChainState::process_stx_unlocks(clarity_tx)?;
+
+        clarity_tx.increment_ustx_liquid_supply(new_unlocked_ustx);
+
+        Ok(lockup_events)
+    }
+
+    /// Called in both follower and miner block assembly paths.
+    ///
+    /// Returns clarity_tx, list of receipts, microblock execution cost,
+    /// microblock fees, microblock burns, list of microblock tx receipts,
+    /// miner rewards tuples, the stacks epoch id, and a boolean that
+    /// represents whether the epoch transition has been applied.
+    ///
+    /// The `burn_dbconn`, `sortition_dbconn`, and `conn` arguments
```

Comment:
I think trait upcasting can still happen though.  You could do something like what I've done for the StackerDB event observer, where you have a "super-trait" that combines `BurnStateDB` and `SortitionDBRef`, as well as a third trait that provided `fn conn(&self) -> &Connection` to recover the inner DB connection.  Then, you could make traits that convert a struct which implements the super-trait into a struct which implements one of the above traits, and create blanket implementations for each of these traits.  For example:

```rust
/// A trait for implementing both mempool event observer methods and stackerdb methods.
/// This is required for event observers to fully report on newly-relayed data.
pub trait RelayEventDispatcher:
    MemPoolEventDispatcher
    + StackerDBEventDispatcher
    + AsMemPoolEventDispatcher
    + AsStackerDBEventDispatcher
{
}
impl<T: MemPoolEventDispatcher + StackerDBEventDispatcher> RelayEventDispatcher for T {}

/// Trait for upcasting to MemPoolEventDispatcher
pub trait AsMemPoolEventDispatcher {
    fn as_mempool_event_dispatcher(&self) -> &dyn MemPoolEventDispatcher;
}

/// Trait for upcasting to StackerDBEventDispatcher
pub trait AsStackerDBEventDispatcher {
    fn as_stackerdb_event_dispatcher(&self) -> &dyn StackerDBEventDispatcher;
}

impl<T: RelayEventDispatcher> AsMemPoolEventDispatcher for T {
    fn as_mempool_event_dispatcher(&self) -> &dyn MemPoolEventDispatcher {
        self
    }
}

impl<T: RelayEventDispatcher> AsStackerDBEventDispatcher for T {
    fn as_stackerdb_event_dispatcher(&self) -> &dyn StackerDBEventDispatcher {
        self
    }
}
```

Reply:
I think it may be more straightforward to just expand the `SortitionDBRef` trait:

```rust
pub trait SortitionDBRef: BurnStateDB {
    fn get_pox_start_cycle_info(
        &self,
        sortition_id: &SortitionId,
        parent_stacks_block_burn_ht: u64,
        cycle_index: u64,
    ) -> Result<Option<PoxStartCycleInfo>, ChainstateError>;

    /// Return an upcasted dynamic reference for the sortition DB
    fn as_burn_state_db(&self) -> &dyn BurnStateDB;

    /// Return a pointer to the underlying sqlite connection or transaction for
    /// this DB reference
    fn sqlite_conn(&self) -> &Connection;
}
```

Because `SortitionDBRef` already requires any implementor also implement `BurnStateDB`, this doesn't really alter the requirements for the trait.

Reply:
Implemented the above (added `as_burn_state_db` to `SortitionDBRef` trait)